### ISSUE TITLE
unknown handle on repository.

### ISSUE THREAD

**Issue by ihulsbus** [concourse/concourse#1821]
# Bug Report

I'm regularly encountering an `unknown handle` on the repository resource:
`unknown handle: 479a6ba9-cc14-4d3b-74fe-7785d4ee9144`

The repository is currently configured as following:
```
- name: repository
  type: git
  source:
    branch: master
    password: ((git-pwd))
    uri: https://private.git.server/repo.git
    username: ((git-user))
  check_every: 24h
  webhook_token: <webhook_token>
```
Sometimes the `unknown handle` goes away after a manual trigger of the webhook, or when re-creating the workers. The webhook does not always work. 

* Concourse version: 3.6.0
* Deployment type (BOSH/Docker/binary): BOSH
* Infrastructure/IaaS: OpenStack
* Browser (if applicable): Any
* Did this used to work? Yes

**Comment by ih16** [comment]
@mvdkleijn FYI

**Comment by EugenMayer** [comment]
As you run on OpenStack i guess your worker / TSA runs on top of a docker-engine, does it?

**Comment by mvdkleijn** [comment]
@EugenMayer Nope. Our deployment is done via Bosh in its own Openstack tenant with the KVM-based VMs.

We we're considering moving from the Bosh release to a Docker-based setup due to some external circumstances. If that's not a good idea, I'd be happy to know why. (however, that's separate from this issue)

I was wondering... could it be that the `check_every: 24h` is causing a too aggressive cleanup of the handle (which I believe is a volume if I'm not mistaken)? (we're setting that due to using a webhook)

Please note that we have other pipelines that are similar to this one but don't seem to suffer from this issue. The only real difference I could think of was the `check_every` setting.

@vito I know you've been looking at lot's of `unknown_handle` issues. This is the same but feels different somehow, do you know if the `check_every` might be the issue?

**Comment by EugenMayer** [comment]
@mvdkleijn yet i was not aware that BOSH based environments have any of those issues - though its just the docker-based ones. The issue you have is known to be related in workers-mixed state persistance.

That means, that especially in docker environments, when a worker is `stop`ed or `down` to be started again, volumes like /worker-state are flushed, while the db volume is persistent and holds partial state e.g. for existing cache ( or something like that ). When the worker comes back alive, he could even have a different hostname in docker envs.

so you have to retire the "old" worker properly to book in the new one ( even though the worker is not new ).

To you your case, when you land a worker before stopping it + persist the /worker-state and down/up, you will get this handle errors. When you do not land +  persist the /worker-state and do down/up, you get a "file not found" issues - when you  land + not persist /worker-state, you get 'handle' issues.

The only thing to properly "restart" a worker right now in docker-env is to retire the worker before shutting him down, and then register him on start. The persistance of /worker-state is not allowed at all over restarts due to that bug.

Explained it briefly for you to maybe match this on the BOSH workflow you use. To fix the issue in docker, if you migrate there, i use a `trap` wrapper to retire a worker before `KILL` https://github.com/EugenMayer/docker-image-concourseci-worker-solid https://github.com/EugenMayer/docker-image-concourseci-worker-solid/blob/master/worker_wrapper.sh#L11
This basically wrappers `wrapper` with a simple script but uses `concourse/concourse` as the base image. Its usage is .e.g here https://github.com/EugenMayer/concourseci-server-boilerplate/blob/master/docker-compose.yml#L49

Doing that, i could setup a proper docker-based production stack without any of those issues anymore, i am using this with rancher catalogs.

**Comment by mvdkleijn** [comment]
@EugenMayer Thanks for the additional information, it's certainly interesting and valuable.

What I understand from your description is that this issue happens mostly when you're spinning up / down workers. In our Bosh setup, we're *not* doing that. Unless Bosh is doing that automatically for us but I don't believe so.

Anyway, thanks for the info on Docker. I'll keep it in mind if/when we migrate.

**Comment by EugenMayer** [comment]
@mvdkleijn yes that is why i am suprised you have this issue with BOSH, it did not seem like this kind of dynamic environment where workers come and go. Still your issue is the same. To fix the issue, in BOSH or not, is yet always to retire the worker, delete worker-state and then probably restart the VM, then worker will be restarted and does register itself again. That should work for your BOSH env, but i guess, thats not what you are yet looking for - you are looking for the cause i guess

**Comment by mvdkleijn** [comment]
@EugenMayer Yes, I'm looking for a final fix. The work-around you describe also doesn't seem to be necessary either since (usually :wink: ) triggering the webhook for the resource will be enough to fix the problem, at least temporarily.

**Comment by EugenMayer** [comment]
@mvdkleijn how would you trigger the webhook manually, `fly check-resource` ? Or are you using a specific button in the UI i could not find? or something you get from `fly -t kwdw checklist` ?

The critical point with that for me was, that it never recovered automatically - without me actually interacting - and scanning for "stucked" resources is not as easy as running `fly builds` and check for failed ones - there is nothing for resources for that, right.

**Comment by ih16** [comment]
We can manually trigger the webhook via Curl: 

```
curl -X POST http://<concourse_url>/api/v1/teams/<team>/pipelines/<pipeline>/resources/<resource>/check/webhook?webhook_token=<TOKEN>
```

**Comment by EugenMayer** [comment]
@ih16 thanks, good to know that there is a webhook automatically.

**Comment by ih16** [comment]
@EugenMayer You'll need to configure a webhook_token in your resource that you want to trigger through a webhook token. See https://concourse.ci/configuring-resources.html. Bottom of the page

**Comment by mvdkleijn** [comment]
We've done some testing by enabling and disabling (i.e. commenting out check_every: 24h and webhook_token in pipeline.yml) the webhook setup.

When disabled, the issue (so far) does not occur. It seems to occur only when using the webhook setup for us.

@vito Does this sound familiar or trigger a bell at all?

**Comment by xtremerui** [comment]
@mvdkleijn is this still happening? What version of concourse is it? Thx.

**Comment by OvermindDL1** [comment]
Still happening here on concourse version 4.2.1.

**Comment by OvermindDL1** [comment]
This still happens on 4.2.1 whenever a worker goes down for reasons outside of my control for quite a period of time.

**Comment by pn-santos** [comment]
We've had this ("unknown handle" on git and docker-image resources) happen about 1 day after the upgrade from 3.14.1 to 4.2.1 (not sure what caused it since it was running fine for about 24h with no issues after the upgrade). 

Resetting the workers (retire + delete state) fixed it and it's now been about a week with further problems.

**Comment by OvermindDL1** [comment]
It was a fresh Concourse 4.2.1 install in my case, and it seems to happen when the system is rebooted, so it's more often than I'd prefer.

**Comment by mvdkleijn** [comment]
Perhaps this is also connected to https://github.com/concourse/concourse/issues/1457?

**Comment by vito** [comment]
This should be fixed by #2588 which will be in 5.0.0.
